### PR TITLE
RBF default epsilon too large 2020

### PR DESCRIPTION
**Pull request by ghost**
This PR was prompted by issue #4523 describing instability in RBF interpolation with noisy data and a user-chosen value of epsilon which was too large for the data interpolated, at least it reminded me that I had posted a bug report #2020 some years ago, and since it was never addressed I thought I should fix it.

The code change is in the calculation of the default epsilon for RBF interpolation which seem to have been in error due to a misinterpretation of the documentation on which it was based.  New unit tests illustrate the behavior described in #4523 (while being deterministic), fail with the previous implementation and pass with the new.

### CODE REVIEW

File: scipy/interpolate/rbf.py [hunk 1]
```diff
@@ -47,7 +47,7 @@
 import sys
 
 from numpy import (sqrt, log, asarray, newaxis, all, dot, exp, eye,
-                   float_)
+                   float_, amax, amin, prod)
```
Comment (ev-br) [inline scipy/interpolate/rbf.py]: Better `import numpy as np`, then use `np.amax` etc for new functions.

File: scipy/interpolate/rbf.py [hunk 2]
```diff
@@ -198,7 +198,10 @@ def __init__(self, *args, **kwargs):
         r = self._call_norm(self.xi, self.xi)
         self.epsilon = kwargs.pop('epsilon', None)
         if self.epsilon is None:
-            self.epsilon = r.mean()
+            dim = self.xi.shape[0]
+            ximax = amax(self.xi, axis=1)
+            ximin = amin(self.xi, axis=1)
+            self.epsilon = pow(prod(ximax - ximin)/self.N, 1.0/dim)
```
Comment (ev-br) [inline scipy/interpolate/rbf.py]: np.power instead of `pow`?

Comment (ev-br) [inline scipy/interpolate/rbf.py]: This change deserves a comment, I'd think.

**Comment by ev-br** [comment]
Thanks for the contrubution!

I've left a few minor style comments inline. One other style request is to squash your commits.

A bigger question is --- in #2020 you refer to some documentation, can you give a more specific reference please? I did not manage to find a ppt on the matlabcentral site.

I had only a cursory look at the rbf.py code and it seems it could use a bit of work. Pauli left some specific suggestions in #4523, would you be interested in looking into it? IMO breaking backwards compat is better be done in one go (not necessarily one PR, just one release)

**Comment by jjgreen** [comment]
The ppt is in the zipfile download at Mathworks, top-right link (I would have attached it here, but GH does not support such).

Re the rework of the whole file: agreed, it does need some work, and probably more than the suggested changes.  I'll look at raising a PR after this one (clearly those changes are outside the scope of this PR).

File: scipy/interpolate/rbf.py (hunk 1 above)
Comment (jjgreen) [inline scipy/interpolate/rbf.py]: done

File: scipy/interpolate/rbf.py (hunk 2 above)
Comment (jjgreen) [inline scipy/interpolate/rbf.py]: done

**Comment by jjgreen** [comment]
Requested changes implemented.

**Comment by argriffing** [comment]
I don't have any significant comment but I noticed that
```python
ximax = np.amax(self.xi, axis=1)
ximin = np.amin(self.xi, axis=1)
ximax - ximin
```
could be changed to
```python
np.ptp(self.xi, axis=1)
```

**Comment by argriffing** [comment]
> A bigger question is --- in #2020 you refer to some documentation, can you give a more specific reference please? I did not manage to find a ppt on the matlabcentral site.

@ev-br The reasoning in #2020 seems pretty intuitive to me.  If you want RBF to approximate a function using something like a blend of its values at 'nearby' points, then it makes sense to care more about the expected distances from points to their neighbors than from points to random points, for the purpose of heuristically determining a default value of 'nearby'.

**Comment by ev-br** [comment]
OK, here's a rebased branch: https://github.com/ev-br/scipy/tree/pr/4643
There are a couple of style tweaks (commit message prefix, docstrings in test cases), and I've added a test case from the original bug report with a random seed so that it's deterministic.
Also used xlogy in the thin-plate kernels to avoid RuntimeWarnings in 0*log(0).

@jjgreen can you take over https://github.com/ev-br/scipy/commit/aaa70fc2f0704641a944090d23925357d1789cd8. Then this is PR can be merged with the green button.

**Comment by ev-br** [comment]
And could you also take over 3847fe3825f0185a028f9c22dd9a9090e98756f1 which fixes  https://github.com/scipy/scipy/issues/1885
(These RuntimeWarnings were coming from 0*log0 and are fixed by using special.xlogy, it looks like)

**Comment by jjgreen** [comment]
Could you please let me know when you've finished adding stuff, I just spent a hour futzing around working out how to merge you changes into my branch, now to find out that you have just obsoleted that work.

**Comment by jjgreen** [comment]
The push after commit 443c2d8 is purely to test the clhanges (due to a recent change in the version of Cython required for a build I can no longer test locally).  A rebased commit will follow a sucessful test-run.

**Comment by ev-br** [comment]
Whoa. I was trying to be helpful, and I'm surprised I've obsoleted your work.
Anyway, 443c2d8 is not necessary, having both tests is fine. You can either keep it or remove it. My preference would be to keep both, but it's up to you. 

I tested the branch locally, as soon as TravisCI tests are done, it's ready to go in IMO.

Cython related failures can be removed by `git rebase upstream/master`.

And here's a how to stuff into your branch: https://github.com/scipy/scipy/pull/4500#issuecomment-73437775
(thanks Pauli)

**Comment by jjgreen** [comment]
Apologies for the irritability, I had 2 attempts at merging from `pr/4634` with merge conflicts, then the email about `ev-br@3847fe3`,  before I found out I should have been cherry-picking,

**Comment by jjgreen** [comment]
BTW, my cython issues are that I have cython 0.21, and there are no debs for 0.22 available yet.

**Comment by ev-br** [comment]
(off-topic for this PR) If you have suggestions for https://github.com/scipy/scipy/issues/4334, it'd be welcome.

I'm nearly sure something along the lines of `pip install cython --upgrade` worked for me.

**Comment by rgommers** [comment]
That should work, Cython is pretty easy to install with pip. At least I've never seen it go wrong (apart from connection timeouts).

**Comment by jjgreen** [comment]
[off topic]

Cython _is_ pretty easy to install via pip if you like living dangerously, `sudo pip install ..` downloads stuff off the internet and then immediately runs a script that you have no opportunity to inspect as root.  Maybe that will work out OK :-).  To do it safely you need a `--user` and then the result is in `~/.local/bin/` which you add to your path or symlink the executable to `user/local/bin` and make sure that you keep an eye on the buglists for the package in case something comes up and ....

Repeat with `gem

**Comment by rgommers** [comment]
@jjgreen true, I also wait for Ubuntu packagers in most cases. But once I'm building Scipy, then pretty much anything else in Python land is easy by comparison - so then I tend to avoid OS packaging completely.

**Comment by ev-br** [comment]
Okay, Travisbot builds all passed, let's put this in. Thanks @jjgreen! Looking forward for your next contributions!

For future PRs (no action asked for at the moment), its better to try to preserve commit authorship, and there is no requirement to squash everything into a single commit :-).

**Pull request merged**